src/main/java/de/tum/cit/aet/artemis/lecture/service/LectureService.java

```diff
@@ -317,4 +318,43 @@ private Lecture filterLectureContentForUser(Lecture lecture, User user) {
         lecture.setLectureUnits(lectureUnitsUserIsAllowedToSee);
         return lecture;
     }
+
+    /**
+     * Derives a set of {@link CalendarEventDTO}s from the {@link Lecture}s associated to the given courseId.
+     * <p>
+     * Whether events are included in the result depends on the visibleDate of the given lecture and whether the
+     * logged-in user is a student of the {@link Course})
+     *
+     * @param courseId      the ID of the course
+     * @param userIsStudent indicates whether the logged-in user is a student of the course
+     * @return the set of results
+     */
+    public Set<CalendarEventDTO> getCalendarEventDTOsFromLectures(Long courseId, boolean userIsStudent) {
+        Set<Lecture> lectures = lectureRepository.findAllByCourseIdWhereStartDateOrEndDateIsNotNull(courseId);
+        return lectures.stream().map(lecture -> deriveEvent(lecture, userIsStudent)).flatMap(Optional::stream).collect(Collectors.toSet());
```

Comment:
Optional:
Makes the intent clearer and is a little more efficient.
```suggestion
        return lectures.stream().map(lecture -> deriveEvent(lecture, userIsStudent)).filter(Optional::isPresent).map(Optional::get).collect(Collectors.toSet());
```
Alternatively, since this is all internal code, you could theoretically ditch the `Optional` and do a simple `.filter(Objects::nonNull)` check.